File: python/백준/백트래킹/15666 N과 M (12).py
def back_tracking(arr, now_length, start, num_arr):
    if now_length == M:
        print(*arr)  # 조건을 만족하는 수열 출력
        return

    for idx in range(start, len(num_arr)):  # 현재 인덱스부터 탐색
        arr.append(num_arr[idx])  # 숫자 추가
        back_tracking(arr, now_length + 1, idx, num_arr)  # 재귀 호출
        arr.pop()  # 백트래킹: 숫자 제거

# 입력 처리
N, M = map(int, input().split())
num_arr = sorted(set(map(int, input().split())))  # 중복 제거 후 정렬

# 백트래킹 호출
back_tracking([], 0, 0, num_arr)

File: cpp/백준/백트래킹/15666 N과 M (12).cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <set>
using namespace std;

int N, M;
vector<int> num_arr;
vector<int> arr;

void back_tracking(int now_length, int start) {
    if (now_length == M) {
        for (int i = 0; i < M; ++i) {
            cout << arr[i];
            if (i + 1 < M) cout << ' ';
        }
        cout << '\n';
        return;
    }

    for (int idx = start; idx < (int)num_arr.size(); ++idx) {
        arr.push_back(num_arr[idx]);
        back_tracking(now_length + 1, idx);
        arr.pop_back();
    }
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);

    cin >> N >> M;
    vector<int> raw(N);
    for (int i = 0; i < N; ++i) cin >> raw[i];
    set<int> unique_set(raw.begin(), raw.end());
    num_arr.assign(unique_set.begin(), unique_set.end());

    back_tracking(0, 0);
    return 0;
}
